recipes/cppcheck/all/conanfile.py

```diff
@@ -44,16 +43,16 @@ def source(self):
         tools.get(**self.conan_data["sources"][self.version],
                   destination=self._source_subfolder, strip_root=True)
 
+    @functools.lru_cache(1)
     def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
-        self._cmake = CMake(self)
-        self._cmake.definitions["USE_Z3"] = self.options.with_z3
-        self._cmake.definitions["HAVE_RULES"] = self.options.have_rules
-        self._cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
-        self._cmake.definitions["ENABLE_OSS_FUZZ"] = False
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+        cmake = CMake(self)
+        cmake.definitions["USE_Z3"] = self.options.with_z3
+        cmake.definitions["HAVE_RULES"] = self.options.have_rules
+        cmake.definitions["USE_MATCHCOMPILER"] = "Auto"
+        cmake.definitions["ENABLE_OSS_FUZZ"] = False
+        cmake.definitions["FILESDIR"] = os.path.join(self.package_folder, "bin")
```

Comment:
@ericLemanissier I don't understand why setting FILESDIR wouldn't be "a good fit for conan" ?

Reply:
(I tested it locally on my Mac and it works)

Reply:
Ok I think I understand: if I get the binary from conan center, FILESDIR will be an absolute path on the build machine and won't be valid locally.

Reply:
I grepped in the current master of cppcheck for `#ifdef FILESDIR` and found 9 matches.
I think every occurrence needs to be changed to call a helper functions that does the following:
```c++
const char* filesdir(const char* default) {
#ifdef FILESDIR
    return FILESDIR
#else
    if (::getenv("CPPCHECK_FILESDIR")) {
        return ::getenv("CPPCHECK_FILESDIR");
    }
    return default;
#endif
}
```

Imho, best would be to create such a patch for current master and send it upstream for feedback + telling why you want to add it.
Based on the feedback, you can improve the patch + add the patch here.

Reply:
> Another possibility, but only applicable to portable applications (not for applications aimed for linux distributions) would be to get the path of the running executable and build a path on top of that.\nBut such code would be platform dependent.

AFAIU, this is already what cppcheck does when FILESDIR macro is not set, except that it does not work on macos for some reason

Reply:
It does things like the following:
```c++
void Settings::loadCppcheckCfg()
{
    std::string fileName = Path::getPathFromFilename(exename) + "cppcheck.cfg";
#ifdef FILESDIR
    if (Path::fileExists(FILESDIR "/cppcheck.cfg"))
        fileName = FILESDIR "/cppcheck.cfg";
#endif

    std::ifstream fin(fileName);
   // ... (do actual work)
```
and `getPathFromFilename` does the following:
```
std::string Path::getPathFromFilename(const std::string &filename)
{
    const std::size_t pos = filename.find_last_of("\\/");

    if (pos != std::string::npos)
        return filename.substr(0, 1 + pos);

    return "";
}
```

So it `cppcheck.cfg` will be found when you run `cppcheck` as e.g. `~/.conan/data/cppcheck/1.2.3.4/_/_/package/123456/bin/cppcheck`, but not when you run it as `cppcheck` (assuming you have `PATH` set-up).

Reply:
`exename` is initialized as follows on Windows:

```c++
#ifdef _WIN32
#include <windows.h>

static char exename[1024] = {0};
#endif

/**
 * Main function of cppcheck
 *
 * @param argc Passed to CppCheck::parseFromArgs()
 * @param argv Passed to CppCheck::parseFromArgs()
 * @return What CppCheckExecutor::check() returns.
 */
int main(int argc, char* argv[])
{
    // MS Visual C++ memory leak debug tracing
#if defined(_MSC_VER) && defined(_DEBUG)
    _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
#endif

    CppCheckExecutor exec;
#ifdef _WIN32
    GetModuleFileNameA(nullptr, exename, sizeof(exename)/sizeof(exename[0])-1);
    argv[0] = exename;
#endif
```

and as follows on Linux:
```c++
#ifdef __linux__
    // Executing cppcheck in PATH. argv[0] does not contain the path.
    if (mSettings->exename.find_first_of("/\\") == std::string::npos) {
        char buf[PATH_MAX] = {0};
        if (FileLister::fileExists("/proc/self/exe") && readlink("/proc/self/exe", buf, sizeof(buf)-1) > 0)
            mSettings->exename = buf;
    }
#endif
```

So @ericLemanissier is right.
Just don't pas `FILESDIR` to cmake and make sure you put the cfg files in the correct location.

Reply:
Asssuming every use of a `.cfg` file is using `exename`.